### PR TITLE
CL-3264 introduce unique-id property to avfvideosrc

### PR DESCRIPTION
**Pull request by lichtzeichner**

### CODE REVIEW

File: subprojects/gst-plugins-bad/sys/applemedia/avfvideosrc.m [hunk 1]
```diff
@@ -1522,6 +1553,9 @@ static void gst_avf_video_src_set_context (GstElement * element,
     case PROP_DEVICE_INDEX:
       impl.deviceIndex = g_value_get_int (value);
       break;
+    case PROP_UNIQUE_ID:
+      impl.uniqueDeviceId = g_value_dup_string (value);
```
Comment (jurica) [inline subprojects/gst-plugins-bad/sys/applemedia/avfvideosrc.m:1566]: You need to free this again (using g_free()) this pointer is not under ARC, see https://clang.llvm.org/docs/AutomaticReferenceCounting.html#retainable-object-pointers

File: subprojects/gst-plugins-bad/sys/applemedia/avfdeviceprovider.m [hunk 2]
```diff
@@ -211,8 +210,8 @@ static void gst_avf_device_set_property (GObject * object, guint prop_id,
   device = GST_AVF_DEVICE_CAST (object);
 
   switch (prop_id) {
-    case PROP_DEVICE_INDEX:
-      device->device_index = g_value_get_int (value);
+    case PROP_UNIQUE_ID:
+      device->unique_id = g_value_dup_string (value);
```
Comment (jurica) [inline subprojects/gst-plugins-bad/sys/applemedia/avfdeviceprovider.m:231]: You need to free this again (using g_free()) this pointer is not under ARC, see https://clang.llvm.org/docs/AutomaticReferenceCounting.html#retainable-object-pointers

File: subprojects/gst-plugins-bad/sys/applemedia/avfvideosrc.m [hunk 3]
```diff
@@ -363,19 +371,29 @@ - (BOOL)openDeviceInput
...
       return NO;
     }
-  } else { // deviceIndex takes priority over position and deviceType
-    NSArray *devices = [AVCaptureDevice devicesWithMediaType:mediaType];
-    if (deviceIndex >= [devices count]) {
-      GST_ELEMENT_ERROR (element, RESOURCE, NOT_FOUND,
-                          ("Invalid video capture device index"), (NULL));
-      return NO;
+  } else { 
+    // uniqueDeviceId takes priority over position and deviceType
```
Comment (jurica) [inline subprojects/gst-plugins-bad/sys/applemedia/avfvideosrc.m]: Mind formatting, I think the comment should be in the same line as the else.

File: subprojects/gst-plugins-bad/sys/applemedia/avfdeviceprovider.m [hunk 4]
```diff
@@ -166,18 +170,30 @@ static void gst_avf_device_set_property (GObject * object, guint prop_id,
...
 
   return elem;
 }
 
+static void
+gst_avf_device_finalize (GObject * object)
+{
+  GstAvfDevice *self = GST_AVF_DEVICE (object);
+
+  if(self->unique_id) {
```
Comment (simonfxr) [inline subprojects/gst-plugins-bad/sys/applemedia/avfdeviceprovider.m]: @lichtzeichner Maybe add a space before the condition

File: subprojects/gst-plugins-bad/sys/applemedia/avfvideosrc.m [hunk 5]
```diff
@@ -363,19 +371,28 @@ - (BOOL)openDeviceInput
...
-      GST_ELEMENT_ERROR (element, RESOURCE, NOT_FOUND,
-                          ("Invalid video capture device index"), (NULL));
-      return NO;
+  } else { // uniqueDeviceId takes priority over position and deviceType
+    NSArray *devices = [AVCaptureDevice devicesWithMediaType:AVMediaTypeVideo];
+
+    [devices enumerateObjectsUsingBlock:^(id object, NSUInteger idx, BOOL *stop) {
+        AVCaptureDevice *dev = [devices objectAtIndex:idx];
+        const gchar* uid = [[dev uniqueID] UTF8String];
+        if(strcmp(uniqueDeviceId, uid) == 0) {
```
Comment (simonfxr) [inline subprojects/gst-plugins-bad/sys/applemedia/avfvideosrc.m]: @lichtzeichner Could use `!strcmp` which seems to be the convention.

Comment (simonfxr) [reply subprojects/gst-plugins-bad/sys/applemedia/avfvideosrc.m]: Also an extra space before the if condition

File: subprojects/gst-plugins-bad/sys/applemedia/avfvideosrc.m [hunk 6]
```diff
@@ -1425,6 +1452,13 @@ static void gst_avf_video_src_set_context (GstElement * element,
 static void
 gst_avf_video_src_finalize (GObject * obj)
 {
+  GstAVFVideoSrcImpl *impl = GST_AVF_VIDEO_SRC_IMPL (obj);
+
+  if(impl.uniqueDeviceId) {
```
Comment (simonfxr) [inline subprojects/gst-plugins-bad/sys/applemedia/avfvideosrc.m]: @lichtzeichner Maybe add a space before the condition and also one for the g_free invocation

File: subprojects/gst-plugins-bad/sys/applemedia/avfvideosrc.m [hunk 7]
```diff
@@ -1522,6 +1559,12 @@ static void gst_avf_video_src_set_context (GstElement * element,
     case PROP_DEVICE_INDEX:
       impl.deviceIndex = g_value_get_int (value);
       break;
+    case PROP_UNIQUE_ID:
+      if(impl.uniqueDeviceId) {
```
Comment (simonfxr) [inline subprojects/gst-plugins-bad/sys/applemedia/avfvideosrc.m]: @lichtzeichner Maybe add some spaces here as well

**Review by simonfxr — COMMENTED** [review]
@lichtzeichner Please take a look

**Pull request merged**